### DISCUSSION TITLE
Need a better dependency resolver

### DISCUSSION THREAD

**Discussion by uberhacker** [skycocker/chromebrew#4956] [Ideas]
I thought I would start a new discussion for issue #4338.  We really need a better algorithm.  The current resolver is very lacking.

**Comment by uberhacker** [comment]
Ideally, the resolver should add dependencies in a logical order so building everything from source is possible if none of the dependencies are currently installed.

**Comment by saltedcoffii** [comment]
I've said this before, and I'll say it again, I'm pretty sure the dependency resolver is doing everything exactly backwards. If we could reverse the output of the dependency resolver entirely, it may work. Might be worth testing.

**Comment by satmandu** [comment]
Also, I think we need to list all dependencies in core both in core and non-core packages. Not being able to properly resolve dependencies in between core packages is a problem, especially when upgrading core packages, because it is impossible to tell which packages need to be upgraded in which order to satisfy interdependencies properly.

Case in point: upgrading libffi.

**Comment by uberhacker** [comment]
@saltedcoffii:  It's one thing to say it's done all wrong and another to come up with some actual code for something better. :)  @satmandu: The idea behind leaving off core dependencies is the assumption that these packages will always be installed.  That said, I understand your point, but there are some core packages that could be included as dependencies in most of the packages.  That is an enormous amount of baggage.
